ml-agents/mlagents/trainers/saver/tf_saver.py

```diff
@@ -0,0 +1,145 @@
+import os
+import shutil
+from typing import Optional
+from mlagents_envs.exception import UnityPolicyException
+from mlagents_envs.logging_util import get_logger
+from mlagents.tf_utils import tf
+from mlagents.trainers.saver.saver import BaseSaver
+from mlagents.trainers.tf.model_serialization import export_policy_model
+from mlagents.trainers.settings import TrainerSettings, SerializationSettings
+from mlagents.trainers.policy.tf_policy import TFPolicy
+from mlagents.trainers import __version__
+
+
+logger = get_logger(__name__)
+
+
+class TFSaver(BaseSaver):
+    """
+    Saver class for TensorFlow
+    """
+
+    def __init__(
+        self, trainer_settings: TrainerSettings, model_path: str, load: bool = False
+    ):
+        super().__init__()
+        self.model_path = model_path
+        self.initialize_path = trainer_settings.init_path
+        self._keep_checkpoints = trainer_settings.keep_checkpoints
+        self.load = load
+
+        # Currently only support saving one policy. This is the one to be saved.
+        self.policy: Optional[TFPolicy] = None
+
+    def register(self, module, init_or_load=True):
+        if isinstance(module, TFPolicy):
+            if self.policy is None:
+                self.policy = module
+                self.graph = self.policy.graph
+                self.sess = self.policy.sess
+                with self.graph.as_default():
+                    self.tf_saver = tf.train.Saver(max_to_keep=self._keep_checkpoints)
+            if init_or_load:
+                self.initialize_or_load(module)
```

Comment:
I think I'd prefer the initialize_or_load() calls to be separate. Either that, or make the abstract method internal (`_initialize_or_load`) and make it clear from the register() docstring that _initialize_or_load will be called.

Reply:
Hmm, there's one catch with this logic that is a bit ugly. `initialize_or_load` really should only be called once the Optimizer is created/registered, since only at that time is the entire graph created and ready for loading. I think it works in this case since both the register calls happen in add_policy after the optimizer is created. Maybe we can get away with an `if isinstance(TFOptimizer)` call `_initialize_or_load()`? Or move it to a separate call in the trainer (i.e. called after both `register` calls happen).

Reply:
Can do that. Then in trainer.py it'll be like 
```
self.saver.register(self.policy)
self.saver.initialize_or_load(self.policy)
self.saver.register(self.optimizer)
```
(looks a bit weird not loading the optimizer but it's fine)

`initialize_or_load` is made a public function for ghost trainer. In that case, this function is called explicitly without registering.

Reply:
You wouldn't need the `self.policy` in `initialize_or_load` right? Since the saver object has already saved the policy object internally when `register(self.policy)` was called.

Reply:
`initialize_or_load` takes the module as input argument.

This is mainly for ghost trainer where the policy in GhostTrainer is not part of the actual trainer (and thus not registered) but need to be initialized.

Reply:
Modified this. Now it takes an optional input argument `policy`. Initialize the input policy if specified, otherwise initialize self.policy.
Now it should look better.

Reply:
Ah got it. Seems good if @andrewcoh agrees w.r.t. the ghost trainer